src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java

```diff
@@ -0,0 +1,215 @@
+     * This is done to avoid having Xray match the components, for which already have the results.
+     * Please notice that this method checks only the direct dependencies as they are the only ones to
+     * be changed by a developer.
+     *
+     * @param root - The DependencyTree root node.
+     * @return a graph of non cached component for Xray scan.
+     */
+    public DependencyTree reduceComponents(DependencyTree root) {
+        DependencyTree reducedTree = (DependencyTree) root.clone();
+        for (DependencyTree child : root.getChildren()) {
+            String componentId = child.toString();
+            if (!scanCache.contains(componentId)) {
+                DependencyTree childCopy = (DependencyTree) child.clone();
+                reducedTree.add(childCopy);
+            }
+        }
+        return reducedTree;
+    }
+
+    @Override
+    public Artifact getArtifactSummary(String componentId) {
+        return scanCache.get(componentId);
+    }
+
+    public static boolean isSupportedInXrayVersion(Version xrayVersion) {
+        return xrayVersion.isAtLeast(MINIMAL_XRAY_VERSION_SUPPORTED_FOR_GRAPH_SCAN);
+    }
+
+    private boolean isSupportedInXrayVersion(Xray xrayClient) {
+        try {
+            if (isSupportedInXrayVersion(xrayClient.system().version())) {
+                return true;
+            }
+            log.error("Unsupported JFrog Xray version: Required JFrog Xray version " + MINIMAL_XRAY_VERSION_SUPPORTED_FOR_GRAPH_SCAN + " and above.");
+        } catch (IOException e) {
+            log.error("JFrog Xray Scan failed. Please check your credentials.", e);
+        }
+        return false;
+    }
+
+    /**
+     * Xray scan a graph of components.
+     *
+     * @param xrayClient      - The Xray client.
+     * @param artifactsToScan - The bulk of components to scan.
+     * @param project         - The JFrog platform project-key parameter to be sent to Xray as context.
+     * @throws IOException in case of connection issues.
+     */
+    private void scanComponents(Xray xrayClient, DependencyTree artifactsToScan, String project) throws IOException {
+        try {
+            if (project != null && !project.isEmpty()) {
+                scanComponentsWithContext(xrayClient, artifactsToScan, project);
+            } else {
+                scanComponentsWithoutContext(xrayClient, artifactsToScan);
+            }
+            // Add to cache non-vulnerable direct dependencies in order to mark them as scanned successfully.
+            // This will allow us to avoid unnecessary future scans.
+            for (DependencyTree child : artifactsToScan.getChildren()) {
+                if (!scanCache.contains(child.getComponentId())) {
+                    scanCache.add(new Artifact(new GeneralInfo(child.toString(),"","",""), new HashSet<>(), new HashSet<>()));
+                }
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```

Comment:
Do you know when it will happen? Nevertheless, I think we should probably print it as we do other exceptions

Reply:
It will be thrown when the thread.sleep fails for some reason...(when we wait for Xray's GET response)
I will add it to the method signature.